### ISSUE TITLE
Long shutdown time and intensive HDD usage updating EPG for large IPTV playlists

### ISSUE THREAD

**Issue by basilgello** [xbmc/xbmc#17800]
### Describe the bug

Kodi creates and deletes the `Epg*.db-journal` for every EPG item being persisted. This causes excessive HDD wear and long shutdown times (2-5 minutes) if IPTV playlist and EPG have several thousands of live TV channels and VODs.

### Expected Behavior

The journal files for `Epg*.db` should be processed in memory so shutdown times are instant (~5-8 seconds on the same playlist / EPG pair vs the unmodified setup)

### Actual Behavior

The iotop graph is shown below:

![iotop-excessive-hdd-usage](https://user-images.githubusercontent.com/3913291/80948267-67bcfa80-8dfa-11ea-9dc6-453407414c3c.png)

The shutdown time is very long (2-5 minutes on AMD E-300 dual-core CPU and 5400rpm HDD)

### Possible Fix

I created the script and the systemd unit mounting the `$KODIPROFILE/userdata/Database` as an overlayfs with the lower directory of original `Database` directory and `tmpfs` mount as upper-directory:

[kodi-mount-overlay.sh](https://github.com/xbmc/xbmc/files/4573561/kodi-mount-overlay.sh.txt)
[kodi-mount-overlay.service](https://github.com/xbmc/xbmc/files/4573566/kodi-mount-overlay.service.txt)

With this ugly hack, the shutdown time and disk usage is desirable.

### To Reproduce

1. Generate the test playlist / EPG pair using the script:
[generate-big-test-m3u-epg.sh](https://github.com/xbmc/xbmc/files/4573573/generate-big-test-m3u-epg.sh.txt)

2. Load the M3U and EPG into IPTV Simple Client and let Kodi process the channels and EPG markers
3. Exit Kodi and observe high disk usage running `sudio iotop` and lots of created-and-vanishing `Epg36.db-journal` in `$KODIPROFILE/userdata/Database`

### Debuglog

The debuglog can be found here:

[kodi.log](https://github.com/xbmc/xbmc/files/4573629/kodi.log)

The log ended after I had to issue
``` shell
kill -9 $(pidof kodi-x11)
```

on my test VM.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Debian 10 buster amd64
 - Kodi version: 19.0 trunk

**Comment by basilgello** [comment]
As a possible fix I tried to append the pragmas here:

https://github.com/xbmc/xbmc/blob/6e731c55b741732a63598a2243bbd2e3064d9c3f/xbmc/dbwrappers/sqlitedataset.cpp#L312

with `PRAGMA temp_store = 2` as advised [here](https://www.sqlite.org/tempfiles.html) but it did not fix the problem.

**Comment by phunkyfish** [comment]
On shutdown the right thing to do here is put all pending changes in a transaction and commit it all in one go.

**Comment by ksooo** [comment]
Does this help?

```diff
diff --git a/xbmc/pvr/epg/Epg.cpp b/xbmc/pvr/epg/Epg.cpp
index cacdde3d0a..ea723d6669 100644
--- a/xbmc/pvr/epg/Epg.cpp
+++ b/xbmc/pvr/epg/Epg.cpp
@@ -291,7 +291,7 @@ std::vector<std::shared_ptr<CPVREpgInfoTag>> CPVREpg::GetTags() const
   return m_tags.GetAllTags();
 }
 
-bool CPVREpg::Persist(const std::shared_ptr<CPVREpgDatabase>& database)
+bool CPVREpg::Persist(const std::shared_ptr<CPVREpgDatabase>& database, bool bQueueWrite)
 {
   if (!database)
   {
@@ -327,7 +327,7 @@ bool CPVREpg::Persist(const std::shared_ptr<CPVREpgDatabase>& database)
     m_bUpdateLastScanTime = false;
   }
 
-  bool bRet = database->CommitInsertQueries();
+  bool bRet = bQueueWrite || database->CommitInsertQueries();
 
   database->Unlock();
 
diff --git a/xbmc/pvr/epg/Epg.h b/xbmc/pvr/epg/Epg.h
index a02ca92e9a..e8371de7b5 100644
--- a/xbmc/pvr/epg/Epg.h
+++ b/xbmc/pvr/epg/Epg.h
@@ -205,9 +205,10 @@ namespace PVR
     /*!
      * @brief Persist this table in the given database
      * @param database The database.
+     * @param bQueueWrite Don't execute the query immediately but queue it if true.
      * @return True if the table was persisted, false otherwise.
      */
-    bool Persist(const std::shared_ptr<CPVREpgDatabase>& database);
+    bool Persist(const std::shared_ptr<CPVREpgDatabase>& database, bool bQueueWrite);
 
     /*!
      * @brief Delete this table from the given database
diff --git a/xbmc/pvr/epg/EpgContainer.cpp b/xbmc/pvr/epg/EpgContainer.cpp
index d8dced3e5a..6fb3f64d50 100644
--- a/xbmc/pvr/epg/EpgContainer.cpp
+++ b/xbmc/pvr/epg/EpgContainer.cpp
@@ -276,8 +276,16 @@ bool CPVREpgContainer::PersistAll(unsigned int iMaxTimeslice) const
   m_critSection.unlock();
 
   const std::shared_ptr<CPVREpgDatabase> database = GetEpgDatabase();
+  if (!database)
+  {
+    CLog::LogF(LOGERROR, "No EPG database");
+    return false;
+  }
+
   XbmcThreads::EndTime processTimeslice(iMaxTimeslice);
 
+//  database->Lock();
+
   for (const auto& epg : epgs)
   {
     if (epg.second && epg.second->NeedsSave())
@@ -285,13 +293,16 @@ bool CPVREpgContainer::PersistAll(unsigned int iMaxTimeslice) const
       CLog::Log(LOGNOTICE, "EPG Container: Persisting events for channel '%s'...",
                 epg.second->GetChannelData()->ChannelName().c_str());
 
-      bReturn &= epg.second->Persist(database);
+      bReturn &= epg.second->Persist(database, true);
     }
 
     if (processTimeslice.IsTimePast())
       break;
   }
 
+  database->CommitInsertQueries();
+//  database->Unlock();
+
   return bReturn;
 }
 
diff --git a/xbmc/pvr/epg/EpgDatabase.cpp b/xbmc/pvr/epg/EpgDatabase.cpp
index a833e71701..adad2a54d2 100644
--- a/xbmc/pvr/epg/EpgDatabase.cpp
+++ b/xbmc/pvr/epg/EpgDatabase.cpp
@@ -958,7 +958,7 @@ bool CPVREpgDatabase::GetLastEpgScanTime(int iEpgId, CDateTime* lastScan)
   return bReturn;
 }
 
-bool CPVREpgDatabase::PersistLastEpgScanTime(int iEpgId, const CDateTime& lastScanTime, bool bQueueWrite /* = false */)
+bool CPVREpgDatabase::PersistLastEpgScanTime(int iEpgId, const CDateTime& lastScanTime, bool bQueueWrite)
 {
   CSingleLock lock(m_critSection);
   std::string strQuery = PrepareSQL("REPLACE INTO lastepgscan(idEpg, sLastScan) VALUES (%u, '%s');",
@@ -967,7 +967,7 @@ bool CPVREpgDatabase::PersistLastEpgScanTime(int iEpgId, const CDateTime& lastSc
   return bQueueWrite ? QueueInsertQuery(strQuery) : ExecuteQuery(strQuery);
 }
 
-int CPVREpgDatabase::Persist(const CPVREpg& epg, bool bQueueWrite /* = false */)
+int CPVREpgDatabase::Persist(const CPVREpg& epg, bool bQueueWrite)
 {
   int iReturn(-1);
   std::string strQuery;
diff --git a/xbmc/pvr/epg/EpgDatabase.h b/xbmc/pvr/epg/EpgDatabase.h
index e4e91c194c..253c4d14ce 100644
--- a/xbmc/pvr/epg/EpgDatabase.h
+++ b/xbmc/pvr/epg/EpgDatabase.h
@@ -224,7 +224,7 @@ namespace PVR
      * @param bQueueWrite Don't execute the query immediately but queue it if true.
      * @return True if it was updated successfully, false otherwise.
      */
-    bool PersistLastEpgScanTime(int iEpgId, const CDateTime& lastScanTime, bool bQueueWrite = false);
+    bool PersistLastEpgScanTime(int iEpgId, const CDateTime& lastScanTime, bool bQueueWrite);
 
     /*!
      * @brief Persist an EPG table. It's entries are not persisted.
@@ -232,7 +232,7 @@ namespace PVR
      * @param bQueueWrite Don't execute the query immediately but queue it if true.
      * @return The database ID of this entry or 0 if bSingleUpdate is false and the query was queued.
      */
-    int Persist(const CPVREpg& epg, bool bQueueWrite = false);
+    int Persist(const CPVREpg& epg, bool bQueueWrite);
 
     /*!
      * @brief Erase all EPG tags with the given epg ID and an end time less than the given time.
```

**Comment by basilgello** [comment]
> Does this help? Not sure it, does.

Thanks for the patch! Let me build it and check against the reproducer.

**Comment by emveepee** [comment]
> 
> 
> On shutdown the right thing to do here is put all pending changes in a transaction and commit it all in one go.
> 
> I think @emveepee was going to look at this at some point.

When I was told the Kodi expected use case was that users would be running for a while I lost bit of interest.  The main problem area after explanation was the clear database feature which is where I hoped I could find a a way of not persisting all the data for a few minutes minutes only to have it immediately deleted.  

I am eager to try the patch as I still feel there wass still a problem for short term users who only want to catch up on the news or something like that after a epg update.

At least with changes to pvr.nextnpvr uid it is minutes an not an hour as was originally the case.

**Comment by emveepee** [comment]
The patch was created before @phunkyfish change https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/epg/EpgContainer.cpp#L285  so you many need to change LOGNOTICE to LOGINFO on line 61

**Comment by basilgello** [comment]
> > Does this help? Not sure it, does.
> 
> Thanks for the patch! Let me build it and check against the reproducer.

It works! It takes incredible 2s to shut Kodi down in a test VM from clean state + reproducer!

**Comment by basilgello** [comment]
> > On shutdown the right thing to do here is put all pending changes in a transaction and commit it all in one go.
> > I think @emveepee was going to look at this at some point.
> 
> When I was told the Kodi expected use case was that users would be running for a while I lost bit of interest. The main problem area after explanation was the clear database feature which is where I hoped I could find a a way of not persisting all the data for a few minutes minutes only to have it immediately deleted.
> 
> I am eager to try the patch as I still feel there wass still a problem for short term users who only want to catch up on the news or something like that after a epg update.
> 
> At least with changes to pvr.nextnpvr uid it is minutes an not an hour as was originally the case.

Same big playlists did not prevent my family's Kodi 17.6 from shutting down while 19.0 did :) Anyway, great to see the patch is working and also great for me to write the script before the regression put too much load on old hardware!

**Comment by ksooo** [comment]
To be honest, I'm astonished that the patch actually works - especially that time is reduced that drastically. On my test system, I was not able to measure any difference. Before and after the patch, storing 180 channels with 30 days of EPG took 5 seconds (macOS + i7 + SSD).

So, I really would like to get feedback from some more users.

**Comment by Tolriq** [comment]
Glad to find this issue, as I was doing tons of tests about web security patch I faced problems to shutdown Kodi. I never waited 5 mins, but with the DEMO PVR addon on Windows i7 hexa core / 64Gb ram / SSD stopping Kodi always block on this for more than 1 / 2 minutes after then I give up and kills it :)

Would gladly test a Windows x64 build if there was a fix for that.

**Comment by emveepee** [comment]
> 
> 
> To be honest, I'm astonished that the patch actually works - especially that time is reduced that drastically. On my test system, I was not able to measure any difference. Before and after the patch, storing 180 channels with 30 days of EPG took 5 seconds (macOS + i7 + SSD).
> 
> So, I really would like to get feedback from some more users.

Yes for me it has gone from minutes to seconds for an incremental update.  It still persists before doing the clear database but the time is very acceptable.  Thank you and well doe. 

To properly test this is tricky since you have to be able to make an incremental update once you have EPG13.db it is never going to slow done until the next scheduled hour of program data comes in.  I suggest between test runs you 

delete from epgtags where iEnddate > 1588723270  (the midnight time of the last day of data in  your guide)

This patch is passing this with flying colour.

**Comment by basilgello** [comment]
The 250 channel list was always the breeze. Issues started with 19.0 on the playlist combining live TV channels and VODs. That's why the reproducer's M3U/EPG pair has so many VOD entries. There is still enough room to improve IPTV Simple Client though - VOD videos should be treated as recordings, not live channels. Then the EPG update logic will operate with 250 live channels and EPG tags for VODs will be fetched only when the end user accesses the VOD.

By the way @Tolriq, are there any plans to implement "Play Programme" in place of "Record" button in Yatse's EPG program tile for finished programmes?
Now the user has to access the catchup programmes using Kodi's TV - EPG interface, which is not convenient if a remote like Yatse is used (especially for people not capable to read the small letters from TV screen at a distance!)

May 4, 2020 3:35:12 PM UTC, emveepee <notifications@github.com> написав(-ла):

**Comment by phunkyfish** [comment]
Please keep the issue thread on topic.

**Comment by ksooo** [comment]
Okay, it seems my patch actually solves this issue. Thus, I will PR it tonight.

**Comment by emveepee** [comment]
It's not just the shutdown issue that has improved.   Prior to this patch the incremental update would only upgrade 5 channels or so at a time and it could take one hour to do it in the background.  Now if there are only 2 or 3 shows per channel with 200 channels they are logged being updated in under one second and there is nothing to do on exit

**Comment by mglae** [comment]
> To be honest, I'm astonished that the patch actually works - especially that time is reduced that drastically. On my test system, I was not able to measure any difference. Before and after the patch, storing 180 channels with 30 days of EPG took 5 seconds (macOS + i7 + SSD).
> 
> So, I really would like to get feedback from some more users.

@ksooo Your feeling was right. While testing on a HDD I can trigger the long shutdown time by changing the VDR server in VNSI client and exit kodi. In this [debug log](http://ix.io/2rq2) it takes 1:58 minutes for ~150 channels.

Tried to implement the advice "Use a Transaction" on channel base:
```diff
diff --git a/xbmc/pvr/epg/EpgTagsContainer.cpp b/xbmc/pvr/epg/EpgTagsContainer.cpp
index 9cd5a36222..f057b5ed0b 100644
--- a/xbmc/pvr/epg/EpgTagsContainer.cpp
+++ b/xbmc/pvr/epg/EpgTagsContainer.cpp
@@ -581,6 +581,7 @@ void CPVREpgTagsContainer::Persist(bool bCommit)
     CLog::Log(LOGDEBUG, "EPG Tags Container: Updating %d, deleting %d events...",
               m_changedTags.size(), m_deletedTags.size());
 
+    m_database->BeginMultipleExecute();
     for (const auto& tag : m_deletedTags)
       m_database->Delete(*tag.second);
 
@@ -599,6 +600,7 @@ void CPVREpgTagsContainer::Persist(bool bCommit)
 
     if (bCommit)
       m_database->CommitInsertQueries();
+    m_database->CommitMultipleExecute();
 
     m_database->Unlock();
   }
```
Now the persisting time is reduced to i.e. [< 3 seconds](http://ix.io/2rq5).

Can this break anything?

**Comment by basilgello** [comment]
Hi @mglae!

While the merged patch reduced waiting times on testing VM, the real HTPC still has problem unsolved. I can disable overlayfs script there and check the exit of Kodi built with your patch.

**Comment by ksooo** [comment]
@basilgello I don't see how this issue should be related to #17840. Could you please explain.

**Comment by basilgello** [comment]
Kai, I mentioned #17800 not #17840. My HTPC still shows long shutdown times with the patches merged to close #17800. I will try the patch above too to check if that improves situation. The testing VM was fine after your patch though.

**Comment by ksooo** [comment]
Which PVR addon are you using?

**Comment by ksooo** [comment]
And please ensure you are using latest Kodi nightly.

**Comment by basilgello** [comment]
Hi Kai!

> Which PVR addon are you using?

I am using pvr.iptvsimple 6.1.2 with Kodi from Debian built on July 2nd.

I am building Kodi & addons right now and will confirm if problem persists and if subsequent rebuild with the patch proposed by mglae solves it.

**Comment by emveepee** [comment]
The @mglae  patch does work miracles for me with pvr.nextpvr.  It is faster on a "real" daily update and the differences are repeatable.  The first @ksooo  fix decreased the update time to minutes and this takes it to seconds, basically v18 performance.    Again I will emphasize the issue isn't only on exit,  it is on the incremental updates which can freeze normal operations.

For testing pvr.nextpvr has a feature to use either EPG source for genre or DVB genres which cause many database delta changes (probably about 1/3 of the records).   Here is the update thread log with the current code  https://paste.kodi.tv/relowizeru.kodi  and here with the patch https://paste.kodi.tv/ijopovahux.kodi

The numbers speak for themselves.    tl;dr version

Unpatched
2020-07-28 14:01:02.085 T:9172     INFO <general>: EPG Container: Persisting unsaved events...
2020-07-28 14:07:53.301 T:9172     INFO <general>: EPG Container: Persisting events done

Patched
2020-07-28 13:56:02.741 T:17096    INFO <general>: EPG Container: Persisting unsaved events...
2020-07-28 13:56:08.040 T:17096    INFO <general>: EPG Container: Persisting events done

A daily update is typically much faster but the time can still overall for my setup is about 2 minutes not 7.

This is on on NVMe database so it isn't disk speed.

**Comment by emveepee** [comment]
Perhaps the speed test is clearer in the logs for the normal  daily incremental background update done each minute

Current code  10 minutes https://paste.kodi.tv/mucenadito.kodi
With patch 3 minutes to https://paste.kodi.tv/votegefexo.kodi  

These operations are disruptive in that the UI blocks during those activity seconds so the fewer the better.    My "production" server has twice as many stations too.

If I quickly exit after loading the guide here is the shutdown effect.

2020-07-29 09:53:38.464 T:5304     INFO <general>: EPG Container: Persisting unsaved events...
2020-07-29 09:54:07.965 T:5304     INFO <general>: EPG Container: Persisting events done

vs

2020-07-29 09:56:19.141 T:9868     INFO <general>: EPG Container: Persisting unsaved events...
2020-07-29 09:56:22.216 T:9868     INFO <general>: EPG Container: Persisting events done

Daily update are primarily inserts but still well worth the patch.

**Comment by phunkyfish** [comment]
Has the patch been PR’d?

**Comment by emveepee** [comment]
@phunkyfish   have been trying and think successful in proving that this issue should be re-opened.  Shouldn't that be done first instead of linking a PR to a closed issue.  If you and @ksooo still don't feel it is an issue how will a PR ever get approved?

**Comment by basilgello** [comment]
@emveepee even I - the author of original issue - can not reopen it for some reason.

I think new PR with 'Fixes #17800' tag will perfectly attach the patch to discussion.

**Comment by phunkyfish** [comment]
Any PR providing a performance improvement with data backing it up would receive little pushback once the implementation seems without side effects.

**Comment by mglae** [comment]
> Has the patch been PR’d?

No. There are issues with it, still WIP.

@emveepee Can you try [this patch](http://ix.io/2sCC)? It is about eight times faster on the HDD, I'm wondering if there are improvements too on the NVMe

**Comment by ksooo** [comment]
Whatever you do, please double check that the performance increase use not just because the queued queries won't be executed anymore.

Using InsertCommitQuery for DELETE queries looks not right to me. Not sure it even works.

**Comment by emveepee** [comment]
> 
> No. There are issues with it, still WIP.

I knew it was WIP because the patch didn't address the commit flag etc

> @emveepee Can you try [this patch](http://ix.io/2sCC)? It is about eight times faster on the HDD, I'm wondering if there are improvements too on the NVMe

Even faster basically the daily update are all inserted in the first incremental update. https://paste.kodi.tv/gasanidowo.kodi and this is a few hours latter so there will be more changes.  There is some new log information at the start I haven't seen before.

For the feature that invalidates more rows causing more replaces it is just as fast https://paste.kodi.tv/homokewaka.kodi many are still done in the first incremental update.  What is 7 minutes in alpha1 is 4 seconds now.

2020-07-29 09:56:19.141 T:9868     INFO <general>: EPG Container: Persisting unsaved events...
2020-07-29 09:56:22.216 T:9868     INFO <general>: EPG Container: Persisting events done

Impressive

**Comment by emveepee** [comment]
> 
> 
> Whatever you do, please double check that the performance increase use not just because the queued queries won't be executed anymore.

When I rerun Kodi after the update I am not seeing big deltas that I see in the original run which indicates epg.db is being updated. On a followup test run just now there was just one update

[code]
2020-07-29 14:19:07.163 T:22036   DEBUG <general>: EPG Container: Persisting events for channel 'Cartoon Network'...
2020-07-29 14:19:07.163 T:22036   DEBUG <general>: EPG Tags Container: Updating 1, deleting 0 events...
2020-07-29 14:19:07.181 T:22036   DEBUG <general>: EPG Container: 3 queries committed.
2020-07-29 14:19:07.181 T:22036   DEBUG <general>: EPG Container: Persist completed.
2020-07-29 14:20:53.532 T:22036    INFO <general>: EPG Container: Persisting unsaved events...
2020-07-29 14:20:53.533 T:22036    INFO <general>: EPG Container: Persisting events done
2020-07-29 14:20:53.534 T:22036   DEBUG <general>: Thread EPGUpdater 22036 terminating
[/code]

**Comment by basilgello** [comment]
Hi @emveepee, can you please diff the sqlite3 <path/to/Epg/db> .dump before and after processing to make sure there are actual changes?

**Comment by emveepee** [comment]
> 
> Hi @emveepee, can you please diff the sqlite3 <path/to/Epg/db> .dump before and after processing to make sure there are actual changes?

Ok I use the genre change and the full diff is too big for paste but this sample looks correct https://paste.kodi.tv/gexacexamu
you can see subtypes change to 256 with text as expected.

**Comment by ksooo** [comment]
@mglae may I suggest to open a PR for your patch. It is much easier then to explain and discuss your intentions. What you propose is not trivial to understand and definitely needs some explanation from your end.

**Comment by mglae** [comment]
@ksooo Coming soon.

**Comment by emveepee** [comment]
@mglae do you want me to prepare a PR based on the last patch?  I think this is too much a performance  improvement to miss out on for alpha2

**Comment by mglae** [comment]
@emveepee In the meantime I do prefer using BeginMultipleExecute() over QueueInsertQuery() like [this](http://ix.io/2vGb) because of the less complex code..

Likely I can test current master today and then create the PR.

**Comment by phunkyfish** [comment]
Just a note this should be tested on MySQL as well as on SQLite.

**Comment by basilgello** [comment]
@phunkyfish for some reason mysql is ~3 times slower for me so I had to stick with tmpfs + overlayfs + sqlit…

**Comment by phunkyfish** [comment]
You mean with @mglae’s patch?

**Comment by basilgello** [comment]
No, without it… just $.02: mysql is considerably slower on EPG with alpha1.

**Comment by phunkyfish** [comment]
Ok, then if you can test it once the PR is ready let’s see does it help MySQL too.

**Comment by mglae** [comment]
@phunkyfish Cite from the [kodi wiki](https://kodi.wiki/view/Advancedsettings.xml#musicdatabase.2Fvideodatabase):

> Kodi also supports using MySQL for PVR (tvdatabase and epgdatabase) as well as the ADSP database. However, there is little to no practical benefit to using MySQL for those databases, and there has been very little testing with them (expect bugs!). Unless you simply want to experiment, or have a very specific need, you should avoid using MySQL for PVR and ADSP.

Is this still true?

**Comment by phunkyfish** [comment]
I’m not sure but if @basilgello is using it I guess not 😉

**Comment by basilgello** [comment]
I tested MySQL as DB backend preparing logs for this bugreport (before @ksooo merged the patch that closed this issue).
It took longer to complete EPG read than sqlite3 so I restored the sqlite3 databases from backup. Now I am using sqlite3
databases on tmpfs overlay.

I will disable overlay to check the shutdown time using 20200817 build (currently deployed on HTPC) and build the current
revision + @mglae patch to check time difference. However it might take a day or two because I need to complete the
investigation of 18.8 build failure in Debian which is not related to this issue :)

**Comment by emveepee** [comment]
> 
> 
> I’m not sure but if @basilgello is using it I guess not 😉
> 
> Regardless of whether it’s fully supported or not it would be good to verify it works with your PR to allows other to finish support for it in the future.

I will set up a test but I am pretty sure for PVR it is not a good idea not only for speed because the delayed writes could in theory write bad data.   The issue of harmonizing EPG update settings is significant too.

client 1 reads epg
client 2 reads new epg
client 2 exits and writes data immediately
client 1 writes potentially stale data over a period of time.

**Comment by emveepee** [comment]
For my first test I thought I'd try the EPG genre test with my production database which modifies 20% of the database.  It is about twice the size of the test database which I tested here https://github.com/xbmc/xbmc/issues/17800#issuecomment-665208532

The results to persist this was 27 minutes, 

2020-08-30 16:42:22.663 T:15904    INFO <general>: EPG Container: Persisting unsaved events...
2020-08-30 17:09:02.737 T:15904    INFO <general>: EPG Container: Persisting events done

Unfortunately with the patch it is still painful.

2020-08-30 17:23:50.782 T:9496     INFO <general>: EPG Container: Persisting unsaved events...
2020-08-30 17:50:32.285 T:9496     INFO <general>: EPG Container: Persisting events done

What is really odd with mysql is these line where it tries to delete items BETWEEN start and stop times for every item.  Perhaps sqlite does this too and it isn't logged.   Here is a couple of idEpg, there are 35k of these in my log.

```
2020-08-30 17:33:24.046 T:9496    DEBUG <general>: EPG Container: Persisting events for channel 'Sportsnet Ontario HD'...
2020-08-30 17:33:24.046 T:9496    DEBUG <general>: EPG Tags Container: Updating 188, deleting 0 events...
2020-08-30 17:33:24.046 T:9496    DEBUG <general>: Mysql execute: DELETE FROM epgtags  WHERE idEpg = 112 AND iEndTime >= 1598761801 AND iStartTime <= 1598763599
2020-08-30 17:33:24.075 T:9496    DEBUG <general>: Mysql execute: DELETE FROM epgtags  WHERE idEpg = 112 AND iEndTime >= 1598763601 AND iStartTime <= 1598767199
2020-08-30 17:33:24.109 T:9496    DEBUG <general>: Mysql execute: DELETE FROM epgtags  WHERE idEpg = 112 AND iEndTime >= 1598767201 AND iStartTime <= 1598770799
```

**Comment by emveepee** [comment]
A test of a daily update is not quite as simple as replacing a file but I saw no appreciable difference between the two for an incremental daily update

Current logic 
2020-08-31 09:37:44.523 T:1720     INFO <general>: EPG Container: Persisting unsaved events...
2020-08-31 09:38:25.667 T:1720     INFO <general>: EPG Container: Persisting events done

Patched logic
2020-08-31 10:14:33.579 T:16780    INFO <general>: EPG Container: Persisting unsaved events...
2020-08-31 10:15:11.884 T:16780    INFO <general>: EPG Container: Persisting events done

To put this in perspective though with my patched base system still speeds through the test

2020-08-31 10:37:27.166 T:13284    INFO <general>: EPG Container: Persisting unsaved events...
2020-08-31 10:37:33.144 T:13284    INFO <general>: EPG Container: Persisting events done

**Comment by ksooo** [comment]
> In the meantime I do prefer using BeginMultipleExecute() over QueueInsertQuery()

 My experience while experimenting with BeginMultipleExecute() was that it does not help to gain performance as it internally uses transactions (iirc).

**Comment by ksooo** [comment]
This https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/epg/EpgTagsContainer.cpp#L593 should be optimized. These requests should be queued, like the respective insert calls.

**Comment by mglae** [comment]
@emveepee Thank you for testing.

Two thoughts:

1. Good. The patch does not break MySQL.
2. The Wiki is right.

:wink:

**Comment by mglae** [comment]
@ksooo 
> My experience while experimenting with BeginMultipleExecute() was that it does not help to gain performance as it internally uses transactions (iirc).

Yes, starting with BeginMultipleExecute() the queries are queued and committed in one transaction by calling CommitMultipleExecute(). IMO together with locking the database exactly what we need.

**Comment by basilgello** [comment]
Yesterday I unintentionally found another edge case where committing all at once does not work.

The iptvsimple plugin failed to retrieve the M3U from a network server and all channels become stale. I noticed that and turned off Kodi via kodi-send. And again, it took 5 minutes to stop Kodi and Epg/TV journals were created and deleted by PVRManager thread.

I managed to attach GDB and found out that the https://github.com/xbmc/xbmc/blob/6bec0e53a4deea1f7be0f70e99c360cd9e24a614/xbmc/pvr/PVRManager.cpp#L374 invokes `CThread::Join` and multiple calls of `Removing stale channel…" are flooded in the log.

I will try reproducing it to get proper logs

**Comment by phunkyfish** [comment]
Thanks for raising. Can we create a new issue for this please?
